### PR TITLE
[dynamic-igemm] odd image support in asm igemm xdlops

### PR DESCRIPTION
**Pull request by carlushuang**
~~[NOTE] This PR is based on https://github.com/ROCmSoftwarePlatform/MIOpen/pull/508 for some refactoring work in asm igemm. So review/merge that PR first hence can make this PR clearer to review~~ =>https://github.com/ROCmSoftwarePlatform/MIOpen/pull/508 Merged

odd image size support in asm igemm xdlops, fwd/bwd/wrw all supported.
for fwd/bwd, `gemm_n = n*h*w`, and for wrw `gemm_k = n*h*w`.
hence support odd image also means relax the limit of `n`

This is achieved to add padding to merged dimension `h*w` to multiply of `nxb`. No new kernel is added specific for padding, but need some change in original kernel that can be compatible for both `h*w` evenly divided by `nxb` or not.
_Files changed_ the *.inc, *.s file is generated from [iGEMMgen](https://github.com/ROCmSoftwarePlatform/iGEMMgen), and indeed the solver part code need be reviewed.

### CODE REVIEW

**Comment by JehandadKhan** [comment]
@asroy can we prioritize this for  4.0 ?

**Comment by asroy** [comment]
> @asroy can we prioritize this for 4.0 ?

yes

**Comment by asroy** [comment]
@carlushuang What dimensions (Ho, Wo, Ho*Wo, or GemmN, etc...) do you padded for fwd/bwd-data/bwd-wrw?

**Comment by carlushuang** [comment]
Hi @asroy I pad `ho*wo` (or `h_dslice*w_dslice` for bwd) to a multiply of a value. And the stratage is the same for fwd/bwd/wrw

**Comment by carlushuang** [comment]
Hi @atamazov since https://github.com/ROCmSoftwarePlatform/MIOpen/pull/508 is merged, this PR is more clear to review, espacially the solver code.

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp [hunk 1]
```diff
@@ -233,22 +252,70 @@ static std::tuple<bool,        // is suitable kernel found
...
+        return std::make_tuple(true,
+                               cfg.GetKernelName(),
+                               cfg.GetBlockSize(),
+                               integer_divide_ceil(gemm_m, cfg.gemm_m_per_block) *
+                                   integer_divide_ceil(gemm_n, cfg.gemm_n_per_block));
     }
-    if(pConfig != tunables.end())
+
+    // second try, try find if packed image size match
+    for(const auto& cfg : tunables)
```
Comment (shaojiewang) [inline src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:269]: Is it possible to find a suitable tunable configuration by one iteration?

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp [hunk 2]
```diff
@@ -340,144 +326,125 @@ FindImplicitGemmWrwGTCDynamicXdlopsKernel(const ConvolutionContext& ctx)
...
+    int sel_log2_gemm_k_global_splits = 0;
+
+    int num_cu                = 120;
+    std::vector<int> nxb_list = {16, 8, 4, 1};
+    std::vector<int> nxe_list = {0, 1};
 
     // i=log2(gemm_m_per_block*gemm_n_per_block)  to find largest kernel
     // switch l and r to get differnet kernel size like 256*64 or 64*256
-    for(i = 15; i > 7; i--)
+    for(int pack = 0; pack < 2; pack++)
```
Comment (shaojiewang) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:341]: Could you please add a comment here to explain the meaning of `pack` ?

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp [hunk 3]
```diff
@@ -340,144 +326,125 @@ FindImplicitGemmWrwGTCDynamicXdlopsKernel(const ConvolutionContext& ctx)
...
-                                                     gemm_k_per_block,
-                                                     gemm_k_global_split,
-                                                     nxb,
-                                                     nxe);
-                    }
+                    const auto gemm_m_per_block = swap == 0 ? 1 << r : 1 << l;
+                    const auto gemm_n_per_block = swap == 0 ? 1 << l : 1 << r;
 
-                    if(tunable_index < 0 || tunable_index >= tunables.size())
+                    if(gemm_m % gemm_m_per_block != 0 || gemm_n % gemm_n_per_block != 0)
```
Comment (shaojiewang) [inline src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:355]: The second condition here should be
 `if( c  %  (  gemm_n_per_block / (nxe == 0?1:nxe)  )  != 0)
    return false;`

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp [hunk 4]
```diff
@@ -233,22 +252,70 @@ static std::tuple<bool,        // is suitable kernel found
...
     {
+        const auto b = cfg.nxe == 0
+                           ? h_tilda_slice * w_tilda_slice
+                           : ((h_tilda_slice * w_tilda_slice + cfg.nxb - 1) / cfg.nxb) * cfg.nxb;
+        const auto gemm_n_packed = n * b;
+        if(cfg.nxe == 0)
```
Comment (jane-zxy) [inline src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:275]: for packed cases, we can simply skip (cfg.nxe==0) , right?

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 1 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:269]: I would want to, but currently if change into one iteration it's better to re-consider the whole process of find kernel logic. If based on currently design it is hard to achieve without introduce regression

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp (hunk 2 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:341]: pack means two iteration, first normal case, then non-normal case

File: src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp (hunk 4 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_gtc_bwd.cpp:275]: Yes. nxe==0 case no pack is support, can make things more simple

File: src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp (hunk 3 above)
Comment (carlushuang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:355]: Do you mean add this condition? and `return false` ?

Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp:355]: As PR542[https://github.com/ROCmSoftwarePlatform/MIOpen/pull/542](url)

**Review by shaojiewang — DISMISSED** [review]

**Review by asroy — APPROVED** [review]

**Comment by shaojiewang** [comment]
Containing code of #542 , this PR also fixed #540 .

**Pull request merged**